File: src/audit_log.ts
/**
 * Local audit trail for the TS engine — the mirror of
 * `py/kintsugi/audit_log.py`. One NDJSON line per ledger attempt
 * (fingerprint, outcome, check/code, patch rationale, provider, and the
 * token usage + derived cost of the model calls made for that finding),
 * plus a final summary line. No Langfuse, no network: `jq -s` the file
 * into a table, or read it with any JSONL tool.
 *
 * Gated by --audit-log <path> on the CLI; without it, this is a no-op.
 */

import { mkdirSync, closeSync, openSync, writeSync } from 'node:fs';
import { dirname, resolve } from 'node:path';

interface Usage {
  input: number;
  output: number;
}

export class AuditLog {
  private fd: number | null = null;
  private failed = false;

  constructor(
    private path: string | null,
    private costUsd: (input: number, output: number) => number,
  ) {
    if (!path) return;
    try {
      const abs = resolve(path);
      mkdirSync(dirname(abs), { recursive: true });
      this.fd = openSync(abs, 'a');
    } catch (err) {
      this.failed = true;
      console.error(`kintsugi: audit log unavailable (${(err as Error).message})`);
    }
  }

  get active(): boolean {
    return this.fd !== null && !this.failed;
  }

  close(): void {
    if (this.fd !== null) {
      try {
        closeSync(this.fd);
      } catch {
        // ignore
      }
      this.fd = null;
    }
  }

  private write(record: unknown): void {
    if (!this.active) return;
    try {
      writeSync(this.fd!, `${JSON.stringify(record)}\n`);
    } catch {
      // One bad record must not take the run down — drop the stream so the
      // failure is loud once, not on every line.
      this.failed = true;
      this.close();
    }
  }

  attempt(opts: {
    fingerprint: string;
    outcome: string;
    check?: string;
    code?: string;
    rationale?: string;
    provider?: boolean;
    collateral?: string[];
    usage?: Usage;
    runId?: string;
  }): void {
    const usage = opts.usage ?? { input: 0, output: 0 };
    this.write({
      event: 'attempt',
      runId: opts.runId ?? '',
      at: new Date().toISOString(),
      fingerprint: opts.fingerprint,
      outcome: opts.outcome,
      check: opts.check ?? '',
      code: opts.code ?? '',
      patchRationale: opts.rationale ?? '',
      provider: opts.provider ?? false,
      collateral: opts.collateral ?? [],
      usage,
      costUsd: this.costUsd(usage.input, usage.output),
    });
  }

  summary(opts: {
    runId?: string;
    status?: string;
    iterations?: number;
    committed?: number;
    reverted?: number;
    quarantined?: number;
    usage?: Usage;
  }): void {
    const usage = opts.usage ?? { input: 0, output: 0 };
    this.write({
      event: 'summary',
      runId: opts.runId ?? '',
      at: new Date().toISOString(),
      status: opts.status ?? '',
      iterations: opts.iterations ?? 0,
      committed: opts.committed ?? 0,
      reverted: opts.reverted ?? 0,
      quarantined: opts.quarantined ?? 0,
      usage,
      costUsd: this.costUsd(usage.input, usage.output),
    });
  }
}

File: py/kintsugi/audit_log.py
"""Local audit trail — one NDJSON line per ledger attempt, no Langfuse.

The Langfuse tracer mirrors the ledger, but auditing then depends on an
external service and a query round-trip. This module gives the loop the
same story as a flat file: one structured JSON line per attempt
(fingerprint, outcome, check/code, patch rationale, provider, and the
token usage + derived cost of the model calls made for that finding),
plus a final summary line with the run totals. Read it with any JSONL
tool, or `jq -s` it into a table:

    jq -s '.[] | select(.event == "attempt") | [.fingerprint, .outcome, .costUsd]' audit.jsonl

Gated by --audit-log <path> on both CLIs; without the flag it is a
no-op and adds nothing to normal output.
"""

import json
import os
import sys
import time


def _iso_now() -> str:
    return time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime())


class AuditLog:
    """Best-effort NDJSON writer. Never raises; a bad path only stderr-notes."""

    def __init__(self, path: str | None, cost_usd):
        self.path = path
        self._cost = cost_usd
        self._fh = None
        self._failed = False
        if path:
            try:
                os.makedirs(os.path.dirname(os.path.abspath(path)), exist_ok=True)
                self._fh = open(path, "a", encoding="utf-8")
            except OSError as err:
                print(f"kintsugi: audit log unavailable ({err})", file=sys.stderr)
                self._failed = True

    @property
    def active(self) -> bool:
        return self._fh is not None

    def close(self) -> None:
        if self._fh:
            try:
                self._fh.close()
            except OSError:
                pass
            self._fh = None

    def _write(self, record: dict) -> None:
        if not self._fh or self._failed:
            return
        try:
            self._fh.write(json.dumps(record, ensure_ascii=False) + "\n")
            self._fh.flush()
        except (OSError, ValueError):
            # One bad record must not take the run down — drop the stream
            # so the failure is loud once, not on every line.
            self._failed = True
            try:
                self.close()
            except Exception:
                pass

    def attempt(self, *, fingerprint: str, outcome: str, check: str = "", code: str = "",
                rationale: str = "", provider: bool = False, collateral: list = None,
                input_tokens: int = 0, output_tokens: int = 0, run_id: str = "") -> None:
        self._write({
            "event": "attempt",
            "runId": run_id,
            "at": _iso_now(),
            "fingerprint": fingerprint,
            "outcome": outcome,
            "check": check,
            "code": code,
            "patchRationale": rationale,
            "provider": provider,
            "collateral": collateral or [],
            "usage": {"input": input_tokens, "output": output_tokens},
            "costUsd": self._cost(input_tokens, output_tokens),
        })

    def summary(self, *, run_id: str = "", status: str = "", iterations: int = 0,
                committed: int = 0, reverted: int = 0, quarantined: int = 0,
                total_input: int = 0, total_output: int = 0) -> None:
        self._write({
            "event": "summary",
            "runId": run_id,
            "at": _iso_now(),
            "status": status,
            "iterations": iterations,
            "committed": committed,
            "reverted": reverted,
            "quarantined": quarantined,
            "usage": {"input": total_input, "output": total_output},
            "costUsd": self._cost(total_input, total_output),
        })
